### PR TITLE
Make build fail when Project.afterEvaluate is called after configuration phase

### LINKED ISSUE #15873: Make 'Calling afterEvaluate when the project is already evaluated' an error
### Summary
Turn nagging in `DefaultProject.maybeNagDeprecationOfAfterEvaluateAfterProjectIsEvaluated()` into an error.

### Checklist
- [x] Update the upgrade guide
- [x] Check User Manual for dangling references and outdated documentation

### PR DESCRIPTION
**Pull request by donat**
Fixes #15873

### CODE REVIEW

**Comment by donat** [comment]
@bot-gradle test rFM

**Comment by donat** [comment]
@bot-gradle test rFM

**Comment by donat** [comment]
@bot-gradle test rFM

File: subprojects/core/src/integTest/groovy/org/gradle/configuration/ExecuteUserLifecycleListenerBuildOperationIntegrationTest.groovy [hunk 1]
```diff
@@ -375,15 +375,13 @@ class ExecuteUserLifecycleListenerBuildOperationIntegrationTest extends Abstract
         ]
         def addProjectListeners = { String source, String target = null ->
             """
-            ${target == null ? '' : "project('$target') { project ->"}
```
Comment (donat) [inline subprojects/core/src/integTest/groovy/org/gradle/configuration/ExecuteUserLifecycleListenerBuildOperationIntegrationTest.groovy:378]: This calls `DefaultProject.evaluate()` That happens in execution time. It is exactly what this PR disables.

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_6.adoc [hunk 2]
```diff
@@ -120,6 +120,11 @@ A Gradle build is defined by a `settings.gradle(.kts)` file in the current or pa
 Without a settings file, a Gradle build is undefined.
 Executing Gradle in such a directory is now an error.
 
+==== Calling Project.afterEvaluate() after project evaluation is now an error
+
+Gradle 6.x warns users about the wrong behavior and ignores the target action in this scenario.
+Starting from 7.0 the same case will produce an error. Plugins and build scripts should be adjusted to call `afterEvaluate` only at configuration time.
```
Comment (bmuskalla) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_6.adoc:126]: Is there any general advice to give people encountering this? What did happen before? Would the block have been ignored anyway? If possible, I think something actionable would be great to add here

**Review by bmuskalla — APPROVED** [review]
LGTM

File: subprojects/core/src/integTest/groovy/org/gradle/configuration/ExecuteUserLifecycleListenerBuildOperationIntegrationTest.groovy [hunk 3]
```diff
@@ -401,18 +399,15 @@ class ExecuteUserLifecycleListenerBuildOperationIntegrationTest extends Abstract
 
         buildFile << addGradleListeners('root project script')
         buildFile << addProjectListeners('root project script', ':')
-        buildFile << addProjectListeners('root project script', ':sub')
         applyInlinePlugin(buildFile, 'Project', addProjectListeners('root project plugin'))
         applyScript(buildFile, scriptFile)
 
         subBuildFile << addGradleListeners('sub project script')
-        subBuildFile << addProjectListeners('sub project script', ':')
```
Comment (donat) [inline subprojects/core/src/integTest/groovy/org/gradle/configuration/ExecuteUserLifecycleListenerBuildOperationIntegrationTest.groovy:409]: The listener is registered in a subproject for the root. When the configuration phase reaches this point, the root project is already evaluated, leading to the freshly introduced build failure.

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_6.adoc (hunk 2 above)
Comment (donat) [reply subprojects/docs/src/docs/userguide/migration/upgrading_version_6.adoc:126]: The `afterEvaluate` we set to fail now was ignored in Gradle versions <7.0. How about this:

_If you have such a build failure and the related `afterEvaluate` statement is declared in your build sources then you can simply delete it. If `afterEvaluate` is declared in a plugin then report the issue to the plugin maintainers._

Comment (bmuskalla) [reply subprojects/docs/src/docs/userguide/migration/upgrading_version_6.adoc:126]: Perfect!

**Comment by donat** [comment]
@bot-gradle test rFR

**Pull request merged**